planner: support create binding from existing plan (#11628)
pingcap/docs: sql-plan-management.md

### Create a binding

> You can create a binding for a SQL statement according to a SQL hint or a historical execution plan.

Reviewer: ```suggestion
You can create a binding for a SQL statement according to a SQL statement or a historical execution plan.
```

> #### Create a binding according to a SQL hint

Reviewer: ```suggestion
#### Create a binding according to a SQL statement
```

>
> For `PREPARE` / `EXECUTE` statements and for queries executed with binary protocols, you need to create execution plan bindings for the real query statements, not for the `PREPARE` / `EXECUTE` statements.

#### Create a binding according to a historical execution plan

> To fix a SQL statement to use a historical execution plan, you can use `plan_digest` to bind that historical execution plan to the SQL statement, which is more convenient than using a SQL hint to bind it.

Author: ```suggestion
To fix a SQL statement to use a historical execution plan, you can use `plan_digest` to bind that historical execution plan to the SQL statement, which is more convenient than using a SQL statement to bind it.
```

> To fix a SQL statement to use a historical execution plan, you can use `plan_digest` to bind that historical execution plan to the SQL statement, which is more convenient than using a SQL statement to bind it.

Author: ```suggestion
To fix a SQL statement to use a historical execution plan, you can use `plan_digest` to bind that historical execution plan to the SQL statement, which is more convenient than binding it according to a SQL statement.
```

> To fix a SQL statement to use a historical execution plan, you can use `plan_digest` to bind that historical execution plan to the SQL statement, which is more convenient than binding it according to a SQL statement.

Reviewer 2: ```suggestion
To make the execution plan of a SQL statement fixed to a historical execution plan, you can use `plan_digest` to bind that historical execution plan to the SQL statement, which is more convenient than binding it according to a SQL statement.
```

> **Warning:**
>
> Currently, creating a binding according to a historical execution plan is an experimental feature with unknown risks. Do not use it in production environments.

Currently, this feature has the following limitations:

- The feature generates hints according to historical execution plans and uses the generated hints for binding. Because historical execution plans are stored in [Statement Summary Tables](/statement-summary-tables.md), before using this feature, you need to enable the [`tidb_enable_stmt_summary`](/system-variables.md#tidb_enable_stmt_summary-new-in-v304) system variable first.
- Currently, this feature only supports binding historical execution plans in the `statements_summary` and `statements_summary_history` tables of the current TiDB node. If you get a `can't find any plans` error, you can connect to another TiDB node in the cluster and retry the binding.
- Currently, this feature does not work for queries with subqueries, queries that access TiFlash, or queries that join 3 or more tables.

The SQL statement of this binding method is as follows:

```sql
CREATE [GLOBAL | SESSION] BINDING FROM HISTORY USING PLAN DIGEST 'plan_digest';
```

> This statement binds a execution plan to a SQL statement using `plan_digest`. The default scope is SESSION. For the applicable SQL statements, priorities, scopes, and effective conditions of the created bindings, they are the same as that of [bindings created using SQL hints](#create-a-binding-according-to-a-sql-hint).

Reviewer: ```suggestion
This statement binds a execution plan to a SQL statement using `plan_digest`. The default scope is SESSION. For the applicable SQL statements, priorities, scopes, and effective conditions of the created bindings, they are the same as that of [bindings created using SQL hints](#create-a-binding-according-to-a-sql-statement).
```

> This statement binds a execution plan to a SQL statement using `plan_digest`. The default scope is SESSION. For the applicable SQL statements, priorities, scopes, and effective conditions of the created bindings, they are the same as that of [bindings created using SQL hints](#create-a-binding-according-to-a-sql-statement).

Author: ```suggestion
This statement binds a execution plan to a SQL statement using `plan_digest`. The default scope is SESSION. For the applicable SQL statements, priorities, scopes, and effective conditions of the created bindings, they are the same as that of [bindings created using SQL statements](#create-a-binding-according-to-a-sql-statement).
```

> This statement binds an execution plan to a SQL statement using `plan_digest`. The default scope is SESSION. For the applicable SQL statements, priorities, scopes, and effective conditions of the created bindings, they are the same as that of [bindings created according to SQL statements](#create-a-binding-according-to-a-sql-statement).

Reviewer 2: ```suggestion
This statement binds an execution plan to a SQL statement using `plan_digest`. The default scope is SESSION. The applicable SQL statements, priorities, scopes, and effective conditions of the created bindings are the same as that of [bindings created according to SQL statements](#create-a-binding-according-to-a-sql-statement).
```

| update_time | Updating time |
| charset | Character set |
| collation | Ordering rule |

> | source | The way in which a binding is created, including `manual` (created according to a SQL hint), `history` (created according to a historical execution plan), `capture` (captured automatically by TiDB), and `evolve` (evolved automatically by TiDB) |

Author: ```suggestion
| source | The way in which a binding is created, including `manual` (created according to a SQL statement), `history` (created according to a historical execution plan), `capture` (captured automatically by TiDB), and `evolve` (evolved automatically by TiDB) |
```

> | sql_digest | Digest of normalized SQL statement |

Reviewer 2: ```suggestion
| sql_digest | Digest of a normalized SQL statement |
```

> | plan_digest | Digest of execution plan |

Reviewer 2: ```suggestion
| plan_digest | Digest of an execution plan |
```
